migrate_cpp/output_segment.h

```diff
@@ -0,0 +1,86 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_MIGRATE_CPP_OUTPUT_SEGMENT_H_
+#define CARBON_MIGRATE_CPP_OUTPUT_SEGMENT_H_
+
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+#include "clang/AST/ASTTypeTraits.h"
+#include "common/check.h"
+
+namespace Carbon {
+
+// Represents a segment of the output string. `OutputSegment`s come in two
+// flavors: Text and Node. A text segment holds string text that should be used
+// to be added to the output. A node segment holds a node in Clang's AST and
+// indicates that the output associated to that node should be the output
+// segment that the `RewriteBuilder` (defined below) has attached to that AST
+// node.
+//
+// For example, the output for a binary operator node corresponding to the C++
+// code snippet `f() + 3 * 5`, would be the sequence of three output segments:
+//
+//                  {Node(lhs), Text(" + "), Node(rhs)}
+//
+// The left-hand side and right-hand side can then be queried recursively to
+// determine what their output should be.
+class OutputSegment {
+ private:
```

Comment:
As I'm looking through this again, I generally take the declaration order as a stronger rule, I'd avoid putting `private` first.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

Reply:
(even if it makes the static constexpr public)

Reply:
Done.